rpcs3/rpcs3qt/log_frame.cpp

```diff
@@ -430,41 +430,49 @@ void log_frame::UpdateUI()
 {
 	const auto start = steady_clock::now();
 
+	const auto tty = g_fxo->get<tty_t>();
+
 	// Check TTY logs
-	while (const u64 size = std::max<s64>(0, g_tty_size.load() - (m_tty_file ? m_tty_file.pos() : 0)))
+	if (tty && tty->fd && tty->fd.size() && m_tty_size != tty->fd.size())
 	{
-		std::string buf;
-		buf.resize(size);
-		buf.resize(m_tty_file.read(&buf.front(), buf.size()));
+		// save old selection
+		QTextCursor text_cursor{ m_tty->document() };
+		const int sel_pos = text_cursor.position();
+		int sel_start = text_cursor.selectionStart();
+		int sel_end = text_cursor.selectionEnd();
 
-		if (buf.find_first_of('\0') != umax)
+		std::string buf;
+		usz skip_lines = m_tty_lines;
 		{
-			m_tty_file.seek(s64{0} - buf.size(), fs::seek_mode::seek_cur);
-			break;
+			std::lock_guard lock(tty->mtx);
+			const auto pos = tty->fd.pos();
+			buf = tty->fd.to_string();
+			tty->fd.seek(pos);
+			m_tty_size = tty->fd.size();
 		}
 
-		if (!buf.empty() && m_TTYAct->isChecked())
+		if (m_TTYAct->isChecked())
 		{
-			std::stringstream buf_stream;
-			buf_stream.str(buf);
-			std::string buf_line;
-			while (std::getline(buf_stream, buf_line))
+			// Process each line seperated by '\n'
+			for (usz start_line = 0, end_line = buf.find_first_of('\n'); start_line < buf.size();
```

Comment:
is this buf variable representing the whole TTY ?
or is it only representing the part of the TTY that wasn't written to the GUI yet?

Reply:
It used to represent only not printed TTY,  now it's the whole file.

Reply:
This means you are processing the whole file every 10ms.
Can't you just save the last position and start your loop there?

Reply:
Imagine a game logs a line every second or even faster, you'll end up with a super slow log. This affects both the tty and the regular log. So this section has to be as fast as it can possibly be.

Reply:
Fixed.

Reply:
How? Please make extra commits during code review. Noone can really follow what's happening otherwise.
You can always squash or amend once it's been approved

Reply:
It's fixed by reading only tty->fd.size() - m_tty_size characters.